File: Assistant/lib/setting.dart
import '/common.dart';
import '/module.dart';
import '/utility/convert_helper.dart';
import '/utility/storage_path.dart';
import '/utility/storage_helper.dart';
import '/utility/json_type.dart';
import '/utility/json_helper.dart';
import '/utility/application_font_manager.dart';
import '/widget/export.dart';
import '/view/core_task_worker/setting.dart' as core_task_worker;
import '/view/core_task_worker/submission_type.dart' as core_task_worker;
import '/view/core_task_worker/value_expression.dart' as core_task_worker;

// ----------------

class SettingData {
  String                                                         version;
  StyledThemeMode                                                themeMode;
  Boolean                                                        themeColorState;
  Color                                                          themeColorLight;
  Color                                                          themeColorDark;
  Boolean                                                        themeFontState;
  List<StoragePath>                                              themeFontPath;
  Integer                                                        windowSizeWidth;
  Integer                                                        windowSizeHeight;
  ModuleType                                                     forwarderDefaultTarget;
  Boolean                                                        forwarderImmediateJump;
  StoragePath                                                    moduleConfigurationDirectory;
  Map<String, StoragePath>                                       storagePickerHistoryLocation;
  Map<ModuleLauncherCategory, List<ModuleLauncherConfiguration>> moduleLauncher;
  Map<ModuleType, Object>                                        moduleSetting;
  SettingData({
    required this.version,
    required this.themeMode,
    required this.themeColorState,
    required this.themeColorLight,
    required this.themeColorDark,
    required this.themeFontState,
    required this.themeFontPath,
    required this.windowSizeWidth,
    required this.windowSizeHeight,
    required this.forwarderDefaultTarget,
    required this.forwarderImmediateJump,
    required this.moduleConfigurationDirectory,
    required this.storagePickerHistoryLocation,
    required this.moduleLauncher,
    required this.moduleSetting,
  });
}

class SettingState {
  Future<Void> Function(String title, ModuleType type, Object option)?                             handleLaunch;
  Future<Void> Function(List<StoragePath> resource)?                                               handleForward;
  Future<Void> Function(List<String> command)?                                                     handleCommand;
  Future<Void> Function(Uri link)?                                                                 handleLink;
  GlobalKey<NavigatorState>                                                                        applicationNavigatorKey;
  List<String>                                                                                     themeFontFamliy;
  Future<Void> Function()?                                                                         homeShowOnboarding;
  Future<Void> Function()?                                                                         homeShowLauncher;
  Future<Void> Function(GlobalKey key, ModuleLauncherConfiguration configuration, Boolean silent)? homeInsertPage;
  Future<Void> Function(GlobalKey key)?                                                            homeRemovePage;
  List<List<core_task_worker.ValueExpression>>                                                     coreTaskWorkerSubmissionHistory;
  SettingState({
    required this.handleLaunch,
    required this.handleForward,
    required this.handleCommand,
    required this.handleLink,
    required this.applicationNavigatorKey,
    required this.themeFontFamliy,
    required this.homeShowOnboarding,
    required this.homeShowLauncher,
    required this.homeInsertPage,
    required this.homeRemovePage,
    required this.coreTaskWorkerSubmissionHistory,
  });
}

class SettingProvider with ChangeNotifier {

  // #region constructor

  SettingData data;

  SettingState state;

  // ----------------

  SettingProvider(
  ) :
    this.data = SettingProvider._createDefaultData(),
    this.state = SettingProvider._createDefaultState();

  // #endregion

  // #region action

  Future<Void> reset(
  ) async {
    this.data = SettingProvider._createDefaultData();
    return;
  }

  Future<Void> apply(
  ) async {
    this.state.themeFontFamliy.clear();
    for (var index = 0; index < (!this.data.themeFontState ? 0 : this.data.themeFontPath.length); index++) {
      var family = await ApplicationFontManager.instance.loadFile(this.data.themeFontPath[index]);
      if (family != null && !this.state.themeFontFamliy.contains(family)) {
        this.state.themeFontFamliy.add(family);
      }
    }
    this.notifyListeners();
    return;
  }

  // #endregion

  // #region storage

  Future<StoragePath> file(
  ) async {
    return (await StorageHelper.query(.applicationShared)).join('setting.json');
  }

  // ----------------

  Future<Void> load({
    StoragePath? file = null,
  }) async {
    if (file == null) {
      file = await this.file();
    }
    this.data = SettingProvider._parseDataFromJson((await JsonHelper.decodeFile(file)).jsonObject());
    return;
  }

  Future<Void> save({
    StoragePath? file = null,
    Boolean      apply = true,
  }) async {
    if (file == null) {
      file = await this.file();
    }
    if (apply) {
      await this.apply();
    }
    if (!await StorageHelper.existFile(file)) {
      await StorageHelper.createFile(file);
    }
    await JsonHelper.encodeFile(file, SettingProvider._makeDataToJson(this.data));
    return;
  }

  // #endregion

  // #region utility

  static SettingProvider of(
    BuildContext context, {
    Boolean listen = true,
  }) {
    return Provider.of<SettingProvider>(context, listen: listen);
  }

  // ----------------

  static SettingData _createDefaultData(
  ) {
    return .new(
      version: ApplicationInformation.version,
      themeMode: .system,
      themeColorState: false,
      themeColorLight: .new(0xFF6200EE),
      themeColorDark: .new(0xFFBB86FC),
      themeFontState: false,
      themeFontPath: [],
      windowSizeWidth: 480,
      windowSizeHeight: 840,
      forwarderDefaultTarget: .coreResourceShipper,
      forwarderImmediateJump: false,
      moduleConfigurationDirectory: .new(),
      storagePickerHistoryLocation: {},
      moduleLauncher: {
        for (var category in ModuleLauncherCategory.values) ...{
          if (category == .module) ...{
            category: ModuleType.values.map(ModuleHelper.query).where((it) => !it.draft).map((it) => ModuleLauncherConfiguration(
              title: it.name,
              type: it.type,
              option: it.generateDefaultOption(),
            )).toList(),
          },
          if (category != .module) ...{
            category: [],
          },
        },
      },
      moduleSetting: {
        for (var type in ModuleType.values) ...{
          type: ModuleHelper.query(type).generateDefaultSetting(),
        },
      },
    );
  }

  static SettingState _createDefaultState(
  ) {
    return .new(
      handleLaunch: null,
      handleForward: null,
      handleCommand: null,
      handleLink: null,
      applicationNavigatorKey: .new(),
      themeFontFamliy: [],
      homeShowOnboarding: null,
      homeShowLauncher: null,
      homeInsertPage: null,
      homeRemovePage: null,
      coreTaskWorkerSubmissionHistory: core_task_worker.SubmissionType.values.map((value) => <core_task_worker.ValueExpression>[]).toList(),
    );
  }

  // ----------------

  static JsonObject _makeDataToJson(
    SettingData data,
  ) {
    return {
      'version': data.version.selfAlso((it) => assertTest(it == ApplicationInformation.version)),
      'theme_mode': data.themeMode.selfLet((it) => ConvertHelper.makeEnumerationToStringOfSnakeCase(it)),
      'theme_color_state': data.themeColorState,
      'theme_color_light': data.themeColorLight.toARGB32(),
      'theme_color_dark': data.themeColorDark.toARGB32(),
      'theme_font_state': data.themeFontState,
      'theme_font_path': data.themeFontPath.map((it) => it.emit()).toList(),
      'window_size_width': data.windowSizeWidth,
      'window_size_height': data.windowSizeHeight,
      'forwarder_default_target': data.forwarderDefaultTarget.selfLet((it) => ModuleHelper.query(it).identifier),
      'forwarder_immediate_jump': data.forwarderImmediateJump,
      'module_configuration_directory': data.moduleConfigurationDirectory.emit(),
      'storage_picker_history_location': data.storagePickerHistoryLocation.map((key, value) => .new(key, value.emit())),
      'module_launcher': data.moduleLauncher.selfLet((data) => {
        for (var category in ModuleLauncherCategory.values) ...{
          category.name.jsonKey(): data[category]!.map((data) => {
            'title': data.title,
            'type': data.type.selfLet((it) => ModuleHelper.query(it).identifier),
            'option': data.option.selfLet((it) => ModuleHelper.query(data.type).makeOption(it)),
          }).toList(),
        },
      }),
      'module_setting': data.moduleSetting.selfLet((data) => {
        for (var type in ModuleType.values) ...{
          type.name.jsonKey(): data[type]!.selfLet((data) => ModuleHelper.query(type).makeSetting(data)),
        },
      }),
    };
  }

  static SettingData _parseDataFromJson(
    JsonObject json,
  ) {
    return SettingData(
      version: json.jsonIn('version').jsonString().selfAlso((it) => assertTest(it == ApplicationInformation.version)),
      themeMode: json.jsonIn('theme_mode').jsonString().selfLet((it) => ConvertHelper.parseEnumerationFromStringOfSnakeCase(it, StyledThemeMode.values)),
      themeColorState: json.jsonIn('theme_color_state').jsonBoolean(),
      themeColorLight: json.jsonIn('theme_color_light').jsonInteger().selfLet((it) => .new(it)),
      themeColorDark: json.jsonIn('theme_color_dark').jsonInteger().selfLet((it) => .new(it)),
      themeFontState: json.jsonIn('theme_font_state').jsonBoolean(),
      themeFontPath: json.jsonIn('theme_font_path').jsonArray().cast<String>().map((it) => StoragePath.of(it)).toList(),
      windowSizeWidth: json.jsonIn('window_size_width').jsonInteger(),
      windowSizeHeight: json.jsonIn('window_size_height').jsonInteger(),
      forwarderDefaultTarget: json.jsonIn('forwarder_default_target').jsonString().selfLet((it) => ConvertHelper.parseEnumerationFromStringOfSnakeCase(it, ModuleType.values)),
      forwarderImmediateJump: json.jsonIn('forwarder_immediate_jump').jsonBoolean(),
      moduleConfigurationDirectory: json.jsonIn('module_configuration_directory').jsonString().selfLet((it) => StoragePath.of(it)),
      storagePickerHistoryLocation: json.jsonIn('storage_picker_history_location').jsonObject().cast<String, String>().map((key, value) => .new(key, StoragePath.of(value))),
      moduleLauncher: json.jsonIn('module_launcher').jsonObject().selfLet((json) => {
        for (var category in ModuleLauncherCategory.values) ...{
          category: json.jsonIn(category.name.jsonKey()).jsonArray().map((json) => ModuleLauncherConfiguration(
            title: json.jsonIn('title').jsonString(),
            type: json.jsonIn('type').jsonString().selfLet((it) => ConvertHelper.parseEnumerationFromStringOfSnakeCase(it, ModuleType.values)),
            option: json.jsonIn('option').jsonObject().selfLet((it) => ModuleHelper.query(json.jsonIn('type').jsonString().selfLet((it) => ConvertHelper.parseEnumerationFromStringOfSnakeCase(it, ModuleType.values))).parseOption(it)),
          )).toList(),
        },
      }),
      moduleSetting: json.jsonIn('module_setting').jsonObject().selfLet((json) => {
        for (var type in ModuleType.values) ...{
          type: ModuleHelper.query(type).parseSetting(json.jsonIn(type.name.jsonKey()).jsonObject()),
        },
      }),
    );
  }

  // ----------------

  Future<Void> quickSetup(
    StoragePath homeDirectory,
  ) async {
    this.data.selfAlso((data) {
      data.moduleConfigurationDirectory = homeDirectory.join('assistant');
      data.moduleSetting[ModuleType.coreTaskWorker]!.as<core_task_worker.Setting>().selfAlso((data) {
        data.kernel = homeDirectory.join('kernel');
        data.script = homeDirectory.join('script').join('main.js');
        data.argument = [homeDirectory.emit()];
      });
    });
    return;
  }

  // #endregion

}

File: AssistantPlus/Application/Setting.cs
#pragma warning disable 0,
// ReSharper disable

using Twinning.AssistantPlus;
using Twinning.AssistantPlus.Utility;
using Windows.UI.StartScreen;
using Microsoft.UI.Composition.SystemBackdrops;
using Microsoft.UI.Xaml.Media;
using Colors = Microsoft.UI.Colors;

namespace Twinning.AssistantPlus {

	public record SettingData {
		public String                                  Version                      { get; set; } = default!;
		public CustomThemeBackdrop                     ThemeBackdrop                { get; set; } = default!;
		public CustomThemeMode                         ThemeMode                    { get; set; } = default!;
		public Boolean                                 ThemeColorState              { get; set; } = default!;
		public Integer                                 ThemeColorLight              { get; set; } = default!;
		public Integer                                 ThemeColorDark               { get; set; } = default!;
		public Boolean                                 WindowPositionState          { get; set; } = default!;
		public Integer                                 WindowPositionX              { get; set; } = default!;
		public Integer                                 WindowPositionY              { get; set; } = default!;
		public Boolean                                 WindowSizeState              { get; set; } = default!;
		public Integer                                 WindowSizeWidth              { get; set; } = default!;
		public Integer                                 WindowSizeHeight             { get; set; } = default!;
		public Dictionary<String, String>              StoragePickerHistoryLocation { get; set; } = default!;
		public ModuleType                              ForwarderDefaultTarget       { get; set; } = default!;
		public Boolean                                 ForwarderImmediateJump       { get; set; } = default!;
		public String                                  ModuleConfigurationDirectory { get; set; } = default!;
		public ModuleLauncherSetting                   ModuleLauncher               { get; set; } = default!;
		public View.CoreTaskWorker.Setting             CoreTaskWorker               { get; set; } = default!;
		public View.CoreCommandSender.Setting          CoreCommandSender            { get; set; } = default!;
		public View.CoreResourceShipper.Setting        CoreResourceShipper          { get; set; } = default!;
		public View.PopcapAnimationViewer.Setting      PopcapAnimationViewer        { get; set; } = default!;
		public View.PopcapReflectionDescriptor.Setting PopcapReflectionDescriptor   { get; set; } = default!;
		public View.PopcapPackageBuilder.Setting       PopcapPackageBuilder         { get; set; } = default!;
	}

	public record SettingState {
		public CustomThemeBackdrop?                            ThemeBackdrop                   { get; set; } = default!;
		public CustomThemeMode?                                ThemeMode                       { get; set; } = default!;
		public List<List<View.CoreTaskWorker.ValueExpression>> CoreTaskWorkerSubmissionHistory { get; set; } = default!;
	}

	public class SettingProvider {

		#region constructor

		public SettingData Data { get; private set; }

		public SettingState State { get; }

		// ----------------

		public SettingProvider(
		) {
			this.Data = SettingProvider.CreateDefaultData();
			this.State = SettingProvider.CreateDefaultState();
			return;
		}

		#endregion

		#region action

		public async Task Reset(
		) {
			this.Data = SettingProvider.CreateDefaultData();
			return;
		}

		public async Task Apply(
		) {
			// ThemeBackdrop
			if (this.State.ThemeBackdrop != this.Data.ThemeBackdrop && App.Instance.MainWindowIsInitialized) {
				App.Instance.MainWindow.SystemBackdrop = this.Data.ThemeBackdrop switch {
					CustomThemeBackdrop.Solid          => null,
					CustomThemeBackdrop.MicaBase       => new MicaBackdrop() { Kind = MicaKind.Base },
					CustomThemeBackdrop.MicaAlt        => new MicaBackdrop() { Kind = MicaKind.BaseAlt },
					CustomThemeBackdrop.AcrylicDesktop => new DesktopAcrylicBackdrop() {},
					_                                  => throw new UnreachableException(),
				};
				App.Instance.MainWindow.uBackground.Visibility = this.Data.ThemeBackdrop == CustomThemeBackdrop.Solid ? Visibility.Visible : Visibility.Collapsed;
				this.State.ThemeBackdrop = this.Data.ThemeBackdrop;
			}
			// ThemeMode
			if (this.State.ThemeMode != this.Data.ThemeMode && App.Instance.MainWindowIsInitialized) {
				App.Instance.MainWindow.Content.As<FrameworkElement>().RequestedTheme = this.Data.ThemeMode switch {
					CustomThemeMode.System => ElementTheme.Default,
					CustomThemeMode.Light  => ElementTheme.Light,
					CustomThemeMode.Dark   => ElementTheme.Dark,
					_                      => throw new UnreachableException(),
				};
				App.Instance.MainWindow.AppWindow.TitleBar.ButtonForegroundColor = this.Data.ThemeMode switch {
					CustomThemeMode.System => null,
					CustomThemeMode.Light  => Colors.Black,
					CustomThemeMode.Dark   => Colors.White,
					_                      => throw new UnreachableException(),
				};
				await ControlHelper.IterateDialog(async (it) => {
					it.RequestedTheme = App.Instance.MainWindow.Content.As<FrameworkElement>().RequestedTheme;
				});
				this.State.ThemeMode = this.Data.ThemeMode;
			}
			// ThemeColor
			{
				var customColorOnLight = ConvertHelper.ParseColorFromInteger(this.Data.ThemeColorLight);
				var customColorOnDark = ConvertHelper.ParseColorFromInteger(this.Data.ThemeColorDark);
				foreach (var resourceKey in new[] { "SystemAccentColorDark1", "SystemAccentColorDark2", "SystemAccentColorDark3" }) {
					if (!this.Data.ThemeColorState) {
						App.Instance.Resources.Remove(resourceKey);
					}
					else {
						App.Instance.Resources[resourceKey] = customColorOnLight;
					}
				}
				foreach (var resourceKey in new[] { "SystemAccentColorLight1", "SystemAccentColorLight2", "SystemAccentColorLight3" }) {
					if (!this.Data.ThemeColorState) {
						App.Instance.Resources.Remove(resourceKey);
					}
					else {
						App.Instance.Resources[resourceKey] = customColorOnDark;
					}
				}
			}
			// ModuleLauncher
			{
				var list = new List<JumpListItem>();
				var generateItem = (ModuleLauncherCategory category, ModuleLauncherConfiguration configuration) => {
					return JumpListItem.CreateWithArguments(ProcessHelper.EncodeCommandString(null, ModuleHelper.GenerateArgument(configuration)), "").SelfAlso((it) => {
						it.Logo = new ("ms-appx:///Asset/Logo.png");
						it.GroupName = category switch {
							ModuleLauncherCategory.Module => "",
							ModuleLauncherCategory.Pinned => "Pinned",
							ModuleLauncherCategory.Recent => "Recent",
							_                             => throw new UnreachableException(),
						};
						it.DisplayName = configuration.Title;
						it.Description = "";
					});
				};
				list.AddRange(App.Instance.Setting.Data.ModuleLauncher.Module.Select((it) => generateItem(ModuleLauncherCategory.Module, it)));
				list.AddRange(App.Instance.Setting.Data.ModuleLauncher.Pinned.Select((it) => generateItem(ModuleLauncherCategory.Pinned, it)));
				list.AddRange(App.Instance.Setting.Data.ModuleLauncher.Recent.Select((it) => generateItem(ModuleLauncherCategory.Recent, it)));
				await ApplicationJumpListManager.Instance.Apply(list);
			}
			// CoreTaskWorker.MessageFont
			{
				App.Instance.Resources["CoreTaskWorker.MessageFont"] = this.Data.CoreTaskWorker.MessageFont.Length == 0 ? FontFamily.XamlAutoFontFamily : new (this.Data.CoreTaskWorker.MessageFont);
			}
			return;
		}

		#endregion

		#region storage

		public String File {
			get {
				return $"{App.Instance.SharedDirectory}/setting.json";
			}
		}

		// ----------------

		public async Task Load(
			String? file = null
		) {
			file ??= this.File;
			this.Data = (await JsonHelper.DeserializeFile<SettingData>(file)).SelfAlso((it) => AssertTest(it.Version == ApplicationInformation.Version));
			return;
		}

		public async Task Save(
			String? file  = null,
			Boolean apply = true
		) {
			file ??= this.File;
			if (apply) {
				await this.Apply();
			}
			if (!await StorageHelper.ExistFile(file)) {
				await StorageHelper.CreateFile(file);
			}
			await JsonHelper.SerializeFile(file, this.Data);
			return;
		}

		#endregion

		#region utility

		private static SettingData CreateDefaultData(
		) {
			return new () {
				Version = ApplicationInformation.Version,
				ThemeBackdrop = CustomThemeBackdrop.MicaBase,
				ThemeMode = CustomThemeMode.System,
				ThemeColorState = false,
				ThemeColorLight = 0xFF0078D4L,
				ThemeColorDark = 0xFF4CC2FFL,
				WindowPositionState = false,
				WindowPositionX = 0,
				WindowPositionY = 0,
				WindowSizeState = true,
				WindowSizeWidth = 480,
				WindowSizeHeight = 840,
				StoragePickerHistoryLocation = [],
				ForwarderDefaultTarget = ModuleType.CoreResourceShipper,
				ForwarderImmediateJump = false,
				ModuleConfigurationDirectory = "",
				ModuleLauncher = new () {
					Module = Enum.GetValues<ModuleType>().Select(ModuleHelper.Query).Select((value) => new ModuleLauncherConfiguration() {
						Title = value.Name,
						Type = value.Type,
						Option = [],
						Command = [
							"-window_size",
							$"{value.StandardSize.Item1}",
							$"{value.StandardSize.Item2}",
						],
					}).ToList(),
					Pinned = [],
					Recent = [],
				},
				CoreTaskWorker = new () {
					Kernel = "",
					Script = "",
					Argument = [],
					AutomaticScroll = true,
					ImmediateLaunch = true,
					MessageFont = "",
				},
				CoreCommandSender = new () {
					ParallelForward = false,
				},
				CoreResourceShipper = new () {
					ParallelForward = false,
					EnableFilter = true,
					EnableBatch = false,
				},
				PopcapAnimationViewer = new () {
					ImmediateSelect = true,
					AutomaticPlay = true,
					RepeatPlay = true,
					KeepSpeed = false,
					ShowBoundary = false,
					ImageFilterRule = "",
					SpriteFilterRule = "",
				},
				PopcapReflectionDescriptor = new () {
				},
				PopcapPackageBuilder = new () {
				},
			};
		}

		private static SettingState CreateDefaultState(
		) {
			return new () {
				ThemeBackdrop = null,
				ThemeMode = null,
				CoreTaskWorkerSubmissionHistory = Enum.GetValues<View.CoreTaskWorker.SubmissionType>().Select((value) => new List<View.CoreTaskWorker.ValueExpression>()).ToList(),
			};
		}

		// ----------------

		public async Task QuickSetup(
			String homeDirectory
		) {
			this.Data.ModuleConfigurationDirectory = $"{homeDirectory}/assistant_plus";
			this.Data.CoreTaskWorker.Kernel = $"{homeDirectory}/kernel";
			this.Data.CoreTaskWorker.Script = $"{homeDirectory}/script/main.js";
			this.Data.CoreTaskWorker.Argument = [$"{homeDirectory}"];
			return;
		}

		#endregion

	}

}
